File: backend/routers/gstin.py
"""
GSTIN Validation Router - Validates GSTIN format and extracts details.
"""

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from typing import Optional
import re
from .deps import get_db, get_current_user
from .models import User

router = APIRouter(prefix="/gstin", tags=["GSTIN"])

STATE_CODE_MAP = {
    '01': 'Jammu & Kashmir', '02': 'Himachal Pradesh', '03': 'Punjab',
    '04': 'Chandigarh', '05': 'Uttarakhand', '06': 'Haryana',
    '07': 'Delhi', '08': 'Rajasthan', '09': 'Uttar Pradesh',
    '10': 'Bihar', '11': 'Sikkim', '12': 'Arunachal Pradesh',
    '13': 'Nagaland', '14': 'Manipur', '15': 'Mizoram',
    '16': 'Tripura', '17': 'Meghalaya', '18': 'Assam',
    '19': 'West Bengal', '20': 'Jharkhand', '21': 'Odisha',
    '22': 'Chhattisgarh', '23': 'Madhya Pradesh', '24': 'Gujarat',
    '25': 'Daman & Diu', '26': 'Dadra & Nagar Haveli', '27': 'Maharashtra',
    '28': 'Andhra Pradesh', '29': 'Karnataka', '30': 'Goa',
    '31': 'Lakshadweep', '32': 'Kerala', '33': 'Tamil Nadu',
    '34': 'Puducherry', '35': 'Andaman & Nicobar', '36': 'Telangana',
    '37': 'Andhra Pradesh (New)', '38': 'Ladakh', '97': 'Other Territory',
}

GSTIN_REGEX = re.compile(r'^[0-9]{2}[A-Z]{5}[0-9]{4}[A-Z]{1}[1-9A-Z]{1}Z[0-9A-Z]{1}$')

PAN_ENTITY_TYPES = {
    'C': 'Company', 'P': 'Individual', 'H': 'HUF',
    'F': 'Firm', 'A': 'AOP', 'T': 'Trust',
    'B': 'BOI', 'L': 'Local Authority', 'J': 'Artificial Juridical Person',
    'G': 'Government',
}


class GSTINValidateRequest(BaseModel):
    gstin: str
    company_name: Optional[str] = None
    lead_id: Optional[str] = None


@router.post("/validate")
async def validate_gstin(
    data: GSTINValidateRequest,
    current_user: User = Depends(get_current_user)
):
    """Validate GSTIN format, extract state/PAN, and cross-check company name."""
    gstin = data.gstin.upper().strip()

    result = {
        "gstin": gstin,
        "valid": False,
        "state_code": "",
        "state_name": "",
        "pan": "",
        "entity_type": "",
        "entity_type_label": "",
        "company_match": None,
        "errors": [],
    }

    if len(gstin) != 15:
        result["errors"].append(f"Must be 15 characters (currently {len(gstin)})")
        return result

    if not GSTIN_REGEX.match(gstin):
        result["errors"].append("Invalid GSTIN format")
        return result

    state_code = gstin[:2]
    state_name = STATE_CODE_MAP.get(state_code)
    if not state_name:
        result["errors"].append(f"Invalid state code: {state_code}")
        return result

    pan = gstin[2:12]
    entity_char = pan[3]
    entity_label = PAN_ENTITY_TYPES.get(entity_char, 'Unknown')

    result.update({
        "valid": True,
        "state_code": state_code,
        "state_name": state_name,
        "pan": pan,
        "entity_type": entity_char,
        "entity_type_label": entity_label,
    })

    # Cross-check company name if provided
    company_name = data.company_name
    if not company_name and data.lead_id:
        db = get_db()
        lead = await db.leads.find_one({"id": data.lead_id}, {"_id": 0, "company": 1})
        if lead:
            company_name = lead.get("company", "")

    if company_name:
        pan_name_chars = pan[:5].upper()
        company_clean = re.sub(r'[^A-Z]', '', company_name.upper())

        if entity_char == 'P':
            result["company_match"] = {
                "match": False,
                "hint": "This is a personal PAN, not a company GSTIN"
            }
        elif entity_char in ('C', 'F', 'H', 'A', 'T'):
            if company_clean.startswith(pan_name_chars):
                result["company_match"] = {
                    "match": True,
                    "hint": f"Company name matches PAN entity: {pan_name_chars}"
                }
            elif company_clean[:3] == pan_name_chars[:3]:
                result["company_match"] = {
                    "match": True,
                    "hint": f"Partial match on entity name"
                }
            else:
                result["company_match"] = {
                    "match": False,
                    "hint": f"PAN entity \"{pan_name_chars}\" doesn't match company \"{company_clean[:5]}\""
                }

    return result

File: frontend/src/utils/gstin.js
/**
 * GSTIN Validation & Extraction Utilities
 * 
 * GSTIN Format: 22AAAAA0000A1Z5 (15 characters)
 * [01-37] State Code | [10 char] PAN | [1 char] Entity | [Z] Default | [1 char] Checksum
 */

const STATE_CODE_MAP = {
  '01': 'Jammu & Kashmir',
  '02': 'Himachal Pradesh',
  '03': 'Punjab',
  '04': 'Chandigarh',
  '05': 'Uttarakhand',
  '06': 'Haryana',
  '07': 'Delhi',
  '08': 'Rajasthan',
  '09': 'Uttar Pradesh',
  '10': 'Bihar',
  '11': 'Sikkim',
  '12': 'Arunachal Pradesh',
  '13': 'Nagaland',
  '14': 'Manipur',
  '15': 'Mizoram',
  '16': 'Tripura',
  '17': 'Meghalaya',
  '18': 'Assam',
  '19': 'West Bengal',
  '20': 'Jharkhand',
  '21': 'Odisha',
  '22': 'Chhattisgarh',
  '23': 'Madhya Pradesh',
  '24': 'Gujarat',
  '25': 'Daman & Diu',
  '26': 'Dadra & Nagar Haveli',
  '27': 'Maharashtra',
  '28': 'Andhra Pradesh',
  '29': 'Karnataka',
  '30': 'Goa',
  '31': 'Lakshadweep',
  '32': 'Kerala',
  '33': 'Tamil Nadu',
  '34': 'Puducherry',
  '35': 'Andaman & Nicobar',
  '36': 'Telangana',
  '37': 'Andhra Pradesh (New)',
  '38': 'Ladakh',
  '97': 'Other Territory',
};

const GSTIN_REGEX = /^[0-9]{2}[A-Z]{5}[0-9]{4}[A-Z]{1}[1-9A-Z]{1}Z[0-9A-Z]{1}$/;

/**
 * Validate GSTIN format and extract details
 * @param {string} gstin 
 * @returns {{ valid: boolean, stateCode: string, stateName: string, pan: string, errors: string[] }}
 */
export function validateGSTIN(gstin) {
  const result = { valid: false, stateCode: '', stateName: '', pan: '', errors: [] };

  if (!gstin) return result;

  const cleaned = gstin.toUpperCase().trim();

  if (cleaned.length !== 15) {
    result.errors.push(`Must be 15 characters (currently ${cleaned.length})`);
    return result;
  }

  if (!GSTIN_REGEX.test(cleaned)) {
    result.errors.push('Invalid GSTIN format');
    return result;
  }

  const stateCode = cleaned.substring(0, 2);
  const stateName = STATE_CODE_MAP[stateCode];

  if (!stateName) {
    result.errors.push(`Invalid state code: ${stateCode}`);
    return result;
  }

  result.valid = true;
  result.stateCode = stateCode;
  result.stateName = stateName;
  result.pan = cleaned.substring(2, 12);

  return result;
}

/**
 * Check if company name partially matches the PAN entity name in GSTIN
 * PAN chars 1-5 relate to entity name (first 5 chars of surname/org name)
 * @param {string} gstin 
 * @param {string} companyName 
 * @returns {{ match: boolean, hint: string }}
 */
export function matchGSTINCompany(gstin, companyName) {
  if (!gstin || !companyName) return { match: false, hint: '' };

  const cleaned = gstin.toUpperCase().trim();
  if (cleaned.length !== 15) return { match: false, hint: '' };

  // PAN is chars 3-12 (0-indexed: 2-11)
  const pan = cleaned.substring(2, 12);
  // 4th char of PAN indicates type: C=Company, P=Person, F=Firm, etc.
  const entityType = pan.charAt(3);
  // Chars 1-5 of PAN: first 5 chars of surname (individual) or entity name (company/firm)
  const panNameChars = pan.substring(0, 4); // First 4 of the 5-char name block

  const companyUpper = companyName.toUpperCase().replace(/[^A-Z]/g, '');

  // For company/firm (C/F/H/A/T), check if company name starts with the PAN name chars
  if (['C', 'F', 'H', 'A', 'T'].includes(entityType)) {
    // Company PAN: first 5 chars = first 5 of company name
    const panName5 = pan.substring(0, 5);
    if (companyUpper.startsWith(panName5)) {
      return { match: true, hint: `Company PAN matches: ${panName5}*` };
    }
    // Partial match (first 3 chars)
    if (companyUpper.substring(0, 3) === panName5.substring(0, 3)) {
      return { match: true, hint: `Partial match on entity name` };
    }
    return { match: false, hint: `PAN entity "${panName5}" doesn't match "${companyUpper.substring(0, 5)}"` };
  }

  // For individual (P), the name block is surname
  if (entityType === 'P') {
    return { match: false, hint: `This is a personal PAN (not a company GSTIN)` };
  }

  return { match: false, hint: '' };
}

export { STATE_CODE_MAP };
